### PR TITLE
Fix blend modes and separate preserve specular

### PR DESCRIPTION
**Pull request by eh-unity**
See: https://github.com/Unity-Technologies/Graphics/pull/5762

### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ]  Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Fix bug case 1260085 (use straight premultiply blend mode)
Separate preserve specular feature from premultiplied blend. Preserve specular is default.
Define URP blends more explicitly (match HDRP) by setting alpha channel blend too and allow pure blend modes.
Added a blend test case.
Added alpha effect to multiply by lerping to white.
Refactored BakedLit and Unlit implementation shader into their own .hlsl files.
Moved universal shader graph template to universal side.

---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 
Added a new test case for blending:
https://github.com/Unity-Technologies/Graphics/blob/8f876cb7beaf56f3ff603e52eb89879e04be39d2/TestProjects/UniversalGraphicsTest/Assets/ReferenceImages/Linear/OSXEditor/Metal/None/066_Blending.png

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers

- Did not account/fix blend modes in fog
- Did not update particle or 2D shaders
- Did not handle accumulate alpha (alpha channel blend with Src == Zero)
- No tests for blend modes & transparencies in fog
- No tests for blending modes with multiple overlapping transparent objects

### CODE REVIEW

**Comment by phi-lira** [comment]
Assigning @ellioman instead of me as reviewer as I have too many PRs under review atm.

**Review by Verasl — CHANGES_REQUESTED** [review]
Seeing keyword changes but no material upgrader is worrying, a material upgrader will be nessesay here, to also add and set values for the new blend properties for alpha.

A material upgrader is a must.

File: com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs [hunk 1]
```diff
@@ -145,6 +149,7 @@ public virtual void FindProperties(MaterialProperty[] properties)
         {
             surfaceTypeProp = FindProperty("_Surface", properties);
             blendModeProp = FindProperty("_Blend", properties);
+            preserveSpecProp = FindProperty("_PreserveSpecular", properties, false);    // TODO: should be mandatory?
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs:152]: Create a discussion about this so we're not leaving a TODO for something that can be decided before this PR lands. :)

File: com.unity.render-pipelines.universal/Shaders/BakedLit.shader [hunk 2]
```diff
@@ -151,7 +157,7 @@ Shader "Universal Render Pipeline/Baked Lit"
                 alpha = OutputAlpha(alpha, _Surface);
 
                 return half4(color, alpha);
-            }
+            }*/
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Shaders/BakedLit.shader:160]: Delete.

File: com.unity.render-pipelines.universal/Shaders/BakedLit.shader [hunk 3]
```diff
@@ -393,7 +402,7 @@ Shader "Universal Render Pipeline/Baked Lit"
...
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Shaders/BakedLit.shader:405]: Delete

File: com.unity.render-pipelines.universal/Shaders/BakedLitForwardPass.hlsl [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    color *= SAMPLE_GI(input.lightmapUV, input.vertexSH, normalWS);
+    #if defined(_SCREEN_SPACE_OCCLUSION)
+        float2 normalizedScreenSpaceUV = GetNormalizedScreenSpaceUV(input.vertex);
+        color *= SampleAmbientOcclusion(normalizedScreenSpaceUV);
+    #endif
+    color = MixFog(color, input.uv0AndFogCoord.z);
+    alpha = OutputAlpha(alpha, _Surface);
+
+    return half4(color, alpha);
+}
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Shaders/BakedLitForwardPass.hlsl:95]: I like moving this into a separate file like we do with the other shaders. 👍

File: com.unity.render-pipelines.universal/Shaders/Unlit.shader [hunk 5]
```diff
@@ -44,67 +47,17 @@ Shader "Universal Render Pipeline/Unlit"
...
-#ifdef _ALPHAPREMULTIPLY_ON
-                color *= alpha;
-#endif
-
-                color = MixFog(color, input.fogCoord);
-
-                return half4(color, alpha);
-            }
+            #include "Packages/com.unity.render-pipelines.universal/Shaders/UnlitInput.hlsl"
+            #include "Packages/com.unity.render-pipelines.universal/Shaders/UnlitForwardPass.hlsl"
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Shaders/Unlit.shader:60]: 👍

**Review by fra-frassineti-unity — CHANGES_REQUESTED** [review]
**NOT APPROVED**:
_"URP/Unlit"_ Shader does not compile on my machine after upgrading materials. For more info see the attached test doc

Version: 2021.1.0a1.495
Revision: trunk eff81b68729d
Built: Wed, 07 Oct 2020 11:32:35 GMT

**Test Doc**:
https://docs.google.com/document/d/1tBreRBnfM6R4cNIWqZS-ixZbO_u_FKHo29nFuNO-Dlw/edit?usp=sharing

**Comment by fra-frassineti-unity** [comment]
Since tomorrow will be my last day at Unity, I re-added @Unity-Technologies/gfx-qa-foundation

**Comment by cinight** [comment]
Talked to Erik. PR is not ready yet. Please add "gfx-qa-foundation" tag as reviewer again when this is ready for testing again.

**Comment by eh-unity** [comment]
Stale. Ended up redoing the change. Link to PR in desc. Closing.

**Pull request closed without merging**